tox.ini

```diff
@@ -0,0 +1,81 @@
+[tox]
+envlist =
+    py{36,37,38}-test{,-alldeps,-devdeps}
+    build_docs
+    linkcheck
+    style
+requires =
+    setuptools >= 30.3.0
+    pip >= 19.3.1
+isolated_build = true
+
+[testenv]
+passenv =
+    HOME
+    WINDIR
+    LC_ALL
+    LC_CTYPE
+    CC
+    CFLAGS
+    CI
+    TRAVIS
+    TEST_READ_HUGE_FILE
+changedir =
+    test: .tmp/{envname}
+    build_docs,linkcheck: docs
+description =
+    test: run tests with pytest
+    build_docs: invoke sphinx-build to build the HTML docs
+    linkcheck: check the links in the HTML docs
+    alldeps: run tests with all dependencies
+    devdeps: run tests with numpy and matplotlib dev versions
+deps =
+    # The oldestdeps tag is intended to be used to install the oldest versions of all
+    # dependencies that have a minimum version
+    oldestdeps: numpy==1.16.*
+    oldestdeps: matplotlib==2.1.*
+    oldestdeps: asdf==2.4.*
+    numpy116: numpy==1.16.*
```

Comment:
These explicit versions seem like a bad idea. Why not just explicitly do this where needed?

Reply:
Because there is no nice way to do this from outside tox (there are ways, but they are ugly). The idea here is to make it so that all test configurations are easy to run either on CI or locally, so that failures in CI can be reproduced locally.

The ``oldestdeps`` is especially useful since there is no other way to make it easy to make sure things always work with the oldest versions of dependencies.

Reply:
Just to give an example, you could pin numpy from outside tox by doing:

```
tox -e test --pypi-filter='numpy==1.16.*'
```

but it's much nicer to just do:

```
tox -e test-numpy116
```

My philosophy with tox and CI is to have as much as possible defined in ``tox.ini`` and as little as possible in the CI config so that we can easily transition between CI services.

Reply:
OK, I see, if this *replaces* things in `.travis.yml`, that is great (and, indeed, better!).